File: nil/cmd/nil/internal/smartaccount/top-up.go
package smartaccount

import (
	"github.com/NilFoundation/nil/nil/cmd/nil/common"
	"github.com/NilFoundation/nil/nil/internal/types"
	"github.com/spf13/cobra"
)

func TopUpCommand(cfg *common.Config) *cobra.Command {
	cmd := &cobra.Command{
		Use:   "top-up [amount] [token-id]",
		Short: "Top up the smart account specified in the config file",
		Args:  cobra.RangeArgs(1, 2),
		RunE: func(cmd *cobra.Command, args []string) error {
			return runTopUp(cmd, args, cfg)
		},
		SilenceUsage: true,
	}

	return cmd
}

func runTopUp(cmd *cobra.Command, args []string, cfg *common.Config) error {
	var amount types.Value
	if err := amount.Set(args[0]); err != nil {
		return err
	}

	var currId string
	if len(args) > 1 {
		currId = args[1]
	}

	return common.RunTopUp(cmd.Context(), "smart account", cfg, cfg.Address, amount, currId, common.Quiet)
}

File: clijs/src/commands/smart-account/top-up.ts
import { Args } from "@oclif/core";
import { BaseCommand } from "../../base.js";
import { bigintArg } from "../../types.js";

export default class SmartAccountTopup extends BaseCommand {
  static override summary = "Top-up address with token";
  static override description = "Top up the smart account specified in the config file";

  static flags = {};

  static args = {
    amount: bigintArg({
      name: "amount",
      required: false,
      description: "The path to the bytecode file",
      default: BigInt("1000000000000000000"),
    }),
    tokenId: Args.string({
      name: "token-id",
      required: false,
      description: "Token Id to top-up",
    }),
  };

  static override examples = ["<%= config.bin %> <%= command.id %>"];

  public async run(): Promise<boolean> {
    const { args } = await this.parse(SmartAccountTopup);

    if (!this.faucetClient) {
      throw new Error("Faucet client is not initialized");
    }
    if (!this.rpcClient) {
      throw new Error("RPC client is not initialized");
    }

    const { smartAccount } = await this.setupSmartAccount();
    const faucets = await this.faucetClient.getAllFaucets();
    const faucetAddress = faucets[args.tokenId ?? "NIL"];
    const txHash = await this.faucetClient.topUpAndWaitUntilCompletion(
      {
        amount: args.amount,
        smartAccountAddress: smartAccount.address,
        faucetAddress: faucetAddress,
      },
      this.rpcClient,
    );
    this.info(`Top-up tx - ${txHash}`);
    return true;
  }
}
